File: tools/check_jcs_vectors.py
#!/usr/bin/env python3
"""Verifica os vectores agnósticos de linguagem RFC 8785."""

import hashlib
import json
import struct
from pathlib import Path

import rfc8785


ROOT = Path(__file__).parent.parent


def main() -> int:
    suite = json.loads((ROOT / "conformance/jcs/vectors.json").read_text(encoding="utf-8"))
    failed = 0
    for vector in suite["vectors"]:
        actual = rfc8785.dumps(vector["value"])
        expected = bytes.fromhex(vector["canonical_utf8_hex"])
        digest = hashlib.sha256(actual).hexdigest()
        ok = actual == expected and digest == vector["sha256"]
        print(("PASS" if ok else "FAIL"), vector["id"])
        failed += not ok
    numbers = json.loads((ROOT / "conformance/jcs/numbers.json").read_text(encoding="utf-8"))
    for ieee_hex, expected_text in numbers["vectors"]:
        value = struct.unpack(">d", bytes.fromhex(ieee_hex))[0]
        ok = rfc8785.dumps(value) == expected_text.encode("ascii")
        print(("PASS" if ok else "FAIL"), "number-" + ieee_hex)
        failed += not ok
    for ieee_hex in numbers["must_reject"]:
        value = struct.unpack(">d", bytes.fromhex(ieee_hex))[0]
        try:
            rfc8785.dumps(value)
            ok = False
        except rfc8785.CanonicalizationError:
            ok = True
        print(("PASS" if ok else "FAIL"), "reject-" + ieee_hex)
        failed += not ok
    return 1 if failed else 0


if __name__ == "__main__":
    raise SystemExit(main())

File: tools/check-jcs-vectors.mjs
#!/usr/bin/env node
// Independent JavaScript checker for the language-neutral RFC 8785 vectors.

import { readFileSync } from "node:fs";
import { createHash } from "node:crypto";

function canonicalize(value) {
  if (value === null || typeof value === "boolean") return JSON.stringify(value);
  if (typeof value === "number") {
    if (!Number.isFinite(value)) throw new Error("non-finite number");
    return JSON.stringify(value);
  }
  if (typeof value === "string") {
    if (/[\uD800-\uDFFF]/u.test(value)) throw new Error("surrogate not permitted");
    return JSON.stringify(value);
  }
  if (Array.isArray(value)) return "[" + value.map(canonicalize).join(",") + "]";
  return "{" + Object.keys(value).sort().map(
    key => JSON.stringify(key) + ":" + canonicalize(value[key])
  ).join(",") + "}";
}

function fromIEEE754(hex) {
  return Buffer.from(hex, "hex").readDoubleBE(0);
}

let failed = 0;
const structural = JSON.parse(readFileSync("conformance/jcs/vectors.json", "utf8"));
for (const vector of structural.vectors) {
  const bytes = Buffer.from(canonicalize(vector.value), "utf8");
  const digest = createHash("sha256").update(bytes).digest("hex");
  const ok = bytes.toString("hex") === vector.canonical_utf8_hex && digest === vector.sha256;
  console.log(ok ? "PASS" : "FAIL", vector.id);
  failed += ok ? 0 : 1;
}

const numbers = JSON.parse(readFileSync("conformance/jcs/numbers.json", "utf8"));
for (const [hex, expected] of numbers.vectors) {
  const ok = canonicalize(fromIEEE754(hex)) === expected;
  console.log(ok ? "PASS" : "FAIL", "number-" + hex);
  failed += ok ? 0 : 1;
}
for (const hex of numbers.must_reject) {
  let ok = false;
  try { canonicalize(fromIEEE754(hex)); } catch { ok = true; }
  console.log(ok ? "PASS" : "FAIL", "reject-" + hex);
  failed += ok ? 0 : 1;
}

process.exitCode = failed ? 1 : 0;
